pkg/orchestrator/planner/compute_forwarder.go

```diff
@@ -0,0 +1,186 @@
+package planner
+
+import (
+	"context"
+
+	"github.com/bacalhau-project/bacalhau/pkg/compute"
+	"github.com/bacalhau-project/bacalhau/pkg/jobstore"
+	"github.com/bacalhau-project/bacalhau/pkg/lib/optional"
+	"github.com/bacalhau-project/bacalhau/pkg/model"
+	"github.com/bacalhau-project/bacalhau/pkg/models"
+	"github.com/bacalhau-project/bacalhau/pkg/orchestrator"
+	"github.com/rs/zerolog/log"
+)
+
+type ComputeForwarder struct {
+	id             string
+	computeService compute.Endpoint
+	jobStore       jobstore.Store
+	evalBroker     orchestrator.EvaluationBroker
+}
+
+type ComputeForwarderParams struct {
+	ID             string
+	ComputeService compute.Endpoint
+	JobStore       jobstore.Store
+}
+
+func NewComputeForwarder(params ComputeForwarderParams) *ComputeForwarder {
+	return &ComputeForwarder{
+		id:             params.ID,
+		computeService: params.ComputeService,
+		jobStore:       params.JobStore,
+	}
+}
+
+func (s *ComputeForwarder) Process(ctx context.Context, plan *models.Plan) error {
+	// notifying nodes is happening in the background, so we can return immediately and not block
+	// the scheduler.
+	// TODO: we need to handle failed notifications. Either enqueue fail the execution and enqueue an evaluation, or
+	//  have a background job that monitors stuck executions and enqueues an evaluation.
+	// TODO: would be nice to notify compute nodes async without blocking the scheduler, but we need to
+	//  serialize the notifications for the same job/execution, which is now done by the evaluation broker.
+	s.doProcess(ctx, plan)
+	return nil
+}
+
+// doProcess performs the actual notifications to compute nodes based on the plan's desired state.
+func (s *ComputeForwarder) doProcess(ctx context.Context, plan *models.Plan) {
+	// TODO: notifying nodes for the same plan can be done in parallel, but we should limit
+	//  the total number of concurrent notifications across plans to avoid overloading the network.
+	for _, exec := range plan.NewExecutions {
+		s.doNotifyAskForBid(ctx, *plan.Job, exec.ID())
+	}
+	for _, u := range plan.UpdatedExecutions {
+		observedState := u.Execution.State
+
+		switch u.DesiredState {
+		case model.ExecutionDesiredStatePending:
+			if observedState == model.ExecutionStateNew {
+				s.doNotifyAskForBid(ctx, *plan.Job, u.Execution.ID())
+			}
+		case model.ExecutionDesiredStateRunning:
+			if observedState == model.ExecutionStateAskForBidAccepted {
+				s.doNotifyBidAccepted(ctx, u.Execution.ID())
+			}
+		case model.ExecutionDesiredStateStopped:
+			if observedState == model.ExecutionStateAskForBidAccepted {
+				s.doNotifyBidRejected(ctx, u.Execution.ID())
+			} else if !observedState.IsTerminal() {
+				s.notifyCancel(ctx, u.Comment, u.Execution.ID())
```

Comment:
I have a TODO statement at lined 38
```
// TODO: we need to handle failed notifications. Either enqueue fail the execution and enqueue an evaluation, or
//  have a background job that monitors stuck executions and enqueues an evaluation.
```

Failing the whole evaluation because we failed to notify one of the compute nodes might be a aggressive. The recommendation is to either have the compute_forwarder enqueue a new evaluation or have it done by a separate background job. A background job can add a lot of value in filling a lot of gaps, but might be slower than fast failure report from the compute forwarder. Maybe we should have both. We need more time to think about this